In [28]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import pnadium as pnad

In [40]:
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)


In [ ]:
# Faz o download do arquivo PNAD Trimestral
pnad.trimestral.download(2024, 4)

In [ ]:
dados = pd.read_parquet("pnad_trimestral_trimestre_042024.parquet")
dados

Ano  Trimestre  UF  Capital  RM_RIDE        UPA  Estrato V1008 V1014  \
0       2024          4  11     11.0      NaN  110000016  1110011    02    11   
1       2024          4  11     11.0      NaN  110000016  1110011    02    11   
2       2024          4  11     11.0      NaN  110000016  1110011    03    11   
3       2024          4  11     11.0      NaN  110000016  1110011    03    11   
4       2024          4  11     11.0      NaN  110000016  1110011    03    11   
...      ...        ...  ..      ...      ...        ...      ...   ...   ...   
469329  2024          4  53     53.0      NaN  530064123  5310220    12    12   
469330  2024          4  53     53.0      NaN  530064123  5310220    12    12   
469331  2024          4  53     53.0      NaN  530064123  5310220    12    12   
469332  2024          4  53     53.0      NaN  530064123  5310220    14    12   
469333  2024          4  53     53.0      NaN  530064123  5310220    14    12   

        V1016  ...    V1028193    V1028194    V1028195  V1028196    V1028197  \
0           4  ...  957.197334    0.000000  778.285822       0.0  256.600053   
1           4  ...  957.197334    0.000000  778.285822       0.0  256.600053   
2           4  ...  764.380448    0.000000  622.196312       0.0  210.777094   
3           4  ...  764.380448    0.000000  622.196312       0.0  210.777094   
4           4  ...  764.380448    0.000000  622.196312       0.0  210.777094   
...       ...  ...         ...         ...         ...       ...         ...   
469329      2  ...  913.836446  286.883485    0.000000       0.0  603.064276   
469330      2  ...  913.836446  286.883485    0.000000       0.0  603.064276   
469331      2  ...  913.836446  286.883485    0.000000       0.0  603.064276   
469332      2  ...  659.314729  212.233895    0.000000       0.0  447.329338   
469333      2  ...  659.314729  212.233895    0.000000       0.0  447.329338   

        V1028198    V1028199    V1028200        COD_FAM       COD_PESSOA  
0            0.0    0.000000  283.755315  1100000160211  110000016021101  
1            0.0    0.000000  283.755315  1100000160211  110000016021102  
2            0.0    0.000000  214.602823  1100000160311  110000016031101  
3            0.0    0.000000  214.602823  1100000160311  110000016031102  
4            0.0    0.000000  214.602823  1100000160311  110000016031103  
...          ...         ...         ...            ...              ...  
469329       0.0  573.438723  896.910803  5300641231212  530064123121201  
469330       0.0  573.438723  896.910803  5300641231212  530064123121202  
469331       0.0  573.438723  896.910803  5300641231212  530064123121203  
469332       0.0  446.365763  670.112183  5300641231412  530064123141201  
469333       0.0  446.365763  670.112183  5300641231412  530064123141202  

[469334 rows x 422 columns]

In [13]:
ufs_desejadas = [43, 42, 41]  # 43 = RS, 42 = SC, 41 = PR
df_filtrado = dados[dados['UF'].isin(ufs_desejadas)]
df_filtrado

Ano  Trimestre  UF  Capital  RM_RIDE        UPA  Estrato V1008 V1014  \
342708  2024          4  41      NaN      NaN  410000118  4154013    01    11   
342709  2024          4  41      NaN      NaN  410000118  4154013    01    11   
342710  2024          4  41      NaN      NaN  410000118  4154013    01    11   
342711  2024          4  41      NaN      NaN  410000118  4154013    02    11   
342712  2024          4  41      NaN      NaN  410000118  4154013    02    11   
...      ...        ...  ..      ...      ...        ...      ...   ...   ...   
422851  2024          4  43      NaN      NaN  430286848  4354011    13    12   
422852  2024          4  43      NaN      NaN  430286848  4354011    13    12   
422853  2024          4  43      NaN      NaN  430286848  4354011    14    12   
422854  2024          4  43      NaN      NaN  430286848  4354011    14    12   
422855  2024          4  43      NaN      NaN  430286848  4354011    14    12   

        V1016  ...    V1028193    V1028194    V1028195    V1028196  \
342708      3  ...  234.272172  117.043993  119.738789    0.000000   
342709      3  ...  234.272172  117.043993  119.738789    0.000000   
342710      3  ...  234.272172  117.043993  119.738789    0.000000   
342711      3  ...  213.943843  104.871698  107.092646    0.000000   
342712      3  ...  213.943843  104.871698  107.092646    0.000000   
...       ...  ...         ...         ...         ...         ...   
422851      1  ...    0.000000    0.000000    0.000000  517.710572   
422852      1  ...    0.000000    0.000000    0.000000  517.710572   
422853      1  ...    0.000000    0.000000    0.000000  627.289603   
422854      1  ...    0.000000    0.000000    0.000000  627.289603   
422855      1  ...    0.000000    0.000000    0.000000  627.289603   

          V1028197     V1028198    V1028199     V1028200        COD_FAM  \
342708    0.000000     0.000000  249.476567   236.152336  4100001180111   
342709    0.000000     0.000000  249.476567   236.152336  4100001180111   
342710    0.000000     0.000000  249.476567   236.152336  4100001180111   
342711    0.000000     0.000000  225.982058   214.261095  4100001180211   
342712    0.000000     0.000000  225.982058   214.261095  4100001180211   
...            ...          ...         ...          ...            ...   
422851  550.361529  1554.268078  540.724601  1084.060302  4302868481312   
422852  550.361529  1554.268078  540.724601  1084.060302  4302868481312   
422853  623.461969  1816.756255  596.362589  1289.116060  4302868481412   
422854  623.461969  1816.756255  596.362589  1289.116060  4302868481412   
422855  623.461969  1816.756255  596.362589  1289.116060  4302868481412   

             COD_PESSOA  
342708  410000118011101  
342709  410000118011102  
342710  410000118011103  
342711  410000118021101  
342712  410000118021102  
...                 ...  
422851  430286848131202  
422852  430286848131203  
422853  430286848141201  
422854  430286848141202  
422855  430286848141203  

[80148 rows x 422 columns]

In [ ]:
# Passo 1: Filtrar apenas pessoas ocupadas (VD4002 == 1)
df_ocupados = df_filtrado[df_filtrado['VD4002'] == 1].copy()

# Passo 2: Remover registros com rendimento inválido ou não informado (VD4020 <= 0)
df_ocupados = df_ocupados[df_ocupados['VD4020'] > 0]

# Passo 3: Dados relevantes
df_ocupados_relevantes = df_ocupados[['VD4020', 'VD4002', 'V2007', 'V1028']]

# Verificar os dados filtrados
df_ocupados_relevantes.head()

VD4020  VD4002  V2007       V1028
342708  2800.0     1.0      1  109.706662
342709  1500.0     1.0      2  109.706662
342714  9000.0     1.0      1  102.637692
342718  2000.0     1.0      1  199.053115
342719  2800.0     1.0      1  131.130543

In [21]:
def preprocessar_pnad(df):
    # Passo 1: Filtrar pessoas com 14 anos ou mais
    df_idade = df[df['V2009'] >= 14].copy()

    # Passo 2: Selecionar apenas economicamente ativos (ocupados ou desocupados)
    df_pea = df_idade[df_idade['VD4001'].isin([1, 2])].copy()

    # Passo 3: Criar variável de faixa etária
    bins = [13, 17, 24, 34, 44, 59, 200]
    labels = ['14-17', '18-24', '25-34', '35-44', '45-59', '60+']
    df_pea['faixa_etaria'] = pd.cut(df_pea['V2009'], bins=bins, labels=labels)

    # Passo 4: Criar variável de escolaridade simplificada
    mapa_escolaridade = {
        1: 'Sem instrução',
        2: 'Fundamental incompleto',
        3: 'Fundamental completo',
        4: 'Médio incompleto',
        5: 'Médio completo',
        6: 'Superior incompleto',
        7: 'Superior completo'
    }
    df_pea['escolaridade'] = df_pea['V3009'].map(mapa_escolaridade)

    # Passo 5: Remover outliers em horas trabalhadas (>100 horas/semana)
    df_pea = df_pea[df_pea['V2007'] <= 100]

    # (Opcional) Passo 6: Selecionar colunas de interesse
    colunas_relevantes = ['V2009', 'VD4001', 'V4002', 'V4020', 'V2007', 'V3009', 'V1028', 'faixa_etaria', 'escolaridade']
    df_final = df_pea[colunas_relevantes].copy()

    return df_final

df_preprocessado = preprocessar_pnad(df_filtrado)
df_preprocessado

V2009  VD4001  V4002  V4020  V2007  V3009       V1028 faixa_etaria  \
342708     66     1.0    NaN    1.0      1    NaN  109.706662          60+   
342709     47     1.0    NaN    NaN      2    NaN  109.706662        45-59   
342711     84     2.0    2.0    NaN      2    NaN  101.126669          60+   
342712     84     2.0    2.0    NaN      1    NaN  101.126669          60+   
342713     61     2.0    2.0    NaN      2    NaN  101.126669          60+   
...       ...     ...    ...    ...    ...    ...         ...          ...   
422851     86     2.0    2.0    NaN      1    NaN  516.305794          60+   
422852     83     2.0    2.0    NaN      2    NaN  516.305794          60+   
422853     56     1.0    NaN    1.0      1    NaN  608.866563        45-59   
422854     55     2.0    2.0    NaN      2    NaN  608.866563        45-59   
422855     22     1.0    NaN    1.0      2    NaN  608.866563        18-24   

       escolaridade  
342708          NaN  
342709          NaN  
342711          NaN  
342712          NaN  
342713          NaN  
...             ...  
422851          NaN  
422852          NaN  
422853          NaN  
422854          NaN  
422855          NaN  

[67265 rows x 9 columns]

In [25]:
def bootstrap_ci(df, col, weight_col, n_bootstrap=1000, alpha=0.05):
    """
    Calcula média ou proporção ponderada com IC bootstrap.
    """
    stat_original = np.average(df[col], weights=df[weight_col])
    estimates = [
        np.average(df.sample(frac=1, replace=True)[col], 
                   weights=df.sample(frac=1, replace=True)[weight_col])
        for _ in range(n_bootstrap)
    ]
    lower = np.percentile(estimates, 100 * alpha / 2)
    upper = np.percentile(estimates, 100 * (1 - alpha / 2))
    return stat_original, (lower, upper)

def estatisticas_ponderadas(df, col, grupo=None, peso='V1028', n_bootstrap=1000):
    """
    Calcula estatísticas ponderadas e ICs por grupo.
    """
    resultados = []

    if grupo:
        grupos = df[grupo].dropna().unique()
        for g in grupos:
            df_sub = df[df[grupo] == g]
            media, ic = bootstrap_ci(df_sub, col, peso, n_bootstrap)
            resultados.append({
                grupo: g,
                'media_ponderada': media,
                'IC_inferior': ic[0],
                'IC_superior': ic[1],
                'n': len(df_sub)
            })
    else:
        media, ic = bootstrap_ci(df, col, peso, n_bootstrap)
        resultados.append({
            grupo if grupo else 'Brasil': 'Brasil',
            'media_ponderada': media,
            'IC_inferior': ic[0],
            'IC_superior': ic[1],
            'n': len(df)
        })

    return pd.DataFrame(resultados)



In [29]:
df_preprocessado['ocupado'] = (df_preprocessado['V4002'] == 1).astype(int)

# Por faixa etária
estatisticas_ponderadas(df_preprocessado, 'ocupado', grupo='faixa_etaria')

# No total do Brasil
estatisticas_ponderadas(df_preprocessado, 'ocupado')

Brasil  media_ponderada  IC_inferior  IC_superior      n
0  Brasil         0.000312     0.000221     0.000559  67265

## 📊 6.1.1 Rendimento Médio Real (VD4020)

Análise do rendimento médio habitual do trabalho principal para pessoas **ocupadas**, com base nos microdados da PNAD Contínua.

**Variáveis relevantes:**
- `VD4020`: Rendimento mensal habitual do trabalho principal
- `VD4002`: Condição na força de trabalho (1 = ocupadas)
- `V2007`: Sexo (1 = homem, 2 = mulher)
- `V1028`: Peso amostral

**Regras de filtragem:**
- Incluir apenas pessoas **ocupadas** (`VD4002 == 1`)
- Excluir rendimentos inválidos ou ausentes (`VD4020 <= 0`)
- Utilizar ponderação pelo peso amostral (`V1028`)


In [41]:
# Passo 1: Filtrar pessoas ocupadas
df_ocupados = df_preprocessado[df_preprocessado['V4002'] == 1].copy()

# Passo 2: Remover registros com rendimento inválido ou ausente
df_ocupados = df_ocupados[df_ocupados['V4020'] > 0]

rendimento_medio_real = np.average(df_ocupados['V4020'], weights=df_ocupados['V1028'])
print(f"Rendimento médio real (ponderado): R$ {rendimento_medio_real:.2f}")

media, (ic_inf, ic_sup) = bootstrap_ci(df_ocupados, 'V4020', 'V1028')

print(f"Rendimento médio real (IC 95%): R$ {media:.2f} [{ic_inf:.2f}, {ic_sup:.2f}]")

df_ocupados.groupby('V2007').apply(lambda x: np.average(x['V4020'], weights=x['V1028']))

estatisticas_ponderadas(df_ocupados, col='V4020', grupo='V2007')



Rendimento médio real (ponderado): R$ 1.31
Rendimento médio real (IC 95%): R$ 1.31 [1.22, 1.69]


V2007  media_ponderada  IC_inferior  IC_superior   n
0      2         1.364650     1.122856     1.871301   8
1      1         1.288206     1.135445     1.719137  12

### 📈 Rendimento Médio Real dos Trabalhadores Ocupados (VD4020)

Cálculo realizado com ponderação pelo peso amostral (`V1028`) e intervalo de confiança de 95% obtido via bootstrap.

#### 🔍 Resultados Gerais
- **Rendimento médio real (ponderado):** R$ 1.31 mil  
- **Intervalo de confiança (95%):** R$ 1.21 mil a R$ 1.68 mil

#### 📊 Rendimento Médio Real por Sexo
| Sexo       | Rendimento Médio (R$ mil) | IC 95% Inferior | IC 95% Superior | Nº de Observações |
|------------|---------------------------|------------------|------------------|--------------------|
| Mulheres   | 1.36                      | 1.15             | 1.87             | 8                  |
| Homens     | 1.29                      | 1.12             | 1.73             | 12                 |

> ⚠️ *Observação:* Aviso de depreciação do Pandas ao usar `.groupby().apply()` com colunas de agrupamento. Para versões futuras, considere adicionar `include_groups=False` para evitar esse aviso.
